### ISSUE TITLE
ActionView::Template::Error (undefined method `downcase' for nil:NilClass)

### ISSUE THREAD

**Issue by ronilaukkarinen** [mastodon/mastodon#29807]
### Steps to reproduce the problem

I'm not sure if it can reproduced.

### Expected behaviour

The user should work.

### Actual behaviour

Something wrong with the user.

### Detailed description

An user from my instance said she can't log in. When I look for her username in the admin it looks like this:

![image](https://github.com/mastodon/mastodon/assets/1534150/eac3d75c-7f69-4aa6-adb6-cba10d9c0828)

When I select it, I get the error page.

The actual error in mastodon-web.service:

```
Mar 29 19:45:05 mementomori bundle[1461]: F, [2024-03-29T19:45:05.874603 #1461] FATAL -- : [e5621941-7639-4c15-901a-a8766cc3b7ec]
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] ActionView::Template::Error (undefined method `downcase' for nil:NilClass):
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     30: = render 'admin/accounts/counters', account: @account
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     31:
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     32: - if @account.local? && @account.user.nil?
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     33:   = link_to t('admin.accounts.unblock_email'), unblock_email_admin_account_path(@account.id), method: :post, class: 'button' if can?(:unblock_email, @account) && CanonicalEmailBlock.matching_account(@account).exists?
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     34: - else
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     35:   .table-wrapper
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]     36:     %table.table.inline-table
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec]
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] app/helpers/email_helper.rb:9:in `email_to_canonical_email'
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] app/helpers/email_helper.rb:16:in `email_to_canonical_email_hash'
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] app/models/canonical_email_block.rb:22:in `block in <class:CanonicalEmailBlock>'
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] app/models/canonical_email_block.rb:23:in `block in <class:CanonicalEmailBlock>'
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] app/views/admin/accounts/show.html.haml:33
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] app/controllers/concerns/localized.rb:11:in `set_locale'
Mar 29 19:45:05 mementomori bundle[1461]: [e5621941-7639-4c15-901a-a8766cc3b7ec] lib/mastodon/rack_middleware.rb:9:in `call'
```

Any idea what happened to this user and why this error occurs? The user is not disabled, suspended, etc. What's going on?

### Mastodon instance

mementomori.social

### Mastodon version

v4.3.0-alpha.3+mementomods-2024-03-22

### Browser name and version

Google Chrome Version 123.0.6312.58 (Official Build) (64-bit)

### Operating system

Pop!_OS by System76 (Ubuntu Jammy 22.04)

**Comment by ronilaukkarinen** [comment]
```bash
RAILS_ENV=production /home/mastodon/live/bin/tootctl accounts refresh username
```

This says OK, but this:

```
RAILS_ENV=production /home/mastodon/live/bin/tootctl accounts modify username --email newemail@gmail.com
```

says: "No user with such username".

**Comment by ronilaukkarinen** [comment]
The account has been created in Jul 30, 2023 and previously I had this in crontab:

```bash
# Remove remote accounts that no longer exist
RAILS_ENV=production /home/mastodon/live/bin/tootctl accounts cull;
```

Could this cause some weird deletion for less used accounts etc.? 

I can't think of any reason why this account is deleted if not deleted by me or the user (waiting her to clarify some things including this).

**Comment by mjankowski** [comment]
Maybe related to https://github.com/mastodon/mastodon/pull/29305

The line the error is from is handling the case in the admin area where there is a local account but not a connected user. Previously the line just did a check for an account with a matching reference_account_id, but after that change it's also looking for matching emails, which for accounts with missing users are going to be nil values.

If that's accurate, there's nothing you or the user did wrong, and this is actually a new-ish edge case to handle after those changes.

**Comment by ronilaukkarinen** [comment]
Ok, thanks. The user replied that she didn't delete the account. So that is still a mystery how the user and her posts got deleted in the first place.

**Comment by ronilaukkarinen** [comment]
I have merged #29305 and while it may fix the issue itself it doesn't repair this issue with the ghost users. Running main from today (2024-05-01) and still getting this for some removed users (no idea how they are removed either):

```log
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] ActionView::Template::Error (undefined method `downcase' for nil:NilClass):
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     30: = render 'admin/accounts/counters', account: @account
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     31:
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     32: - if @account.local? && @account.user.nil?
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     33:   = link_to t('admin.accounts.unblock_email'), unblock_email_admin_account_path(@account.id), method: :post, class: 'button' if can?(:unblock_email, @account) && CanonicalEmailBlock.matching_account(@account).exists?
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     34: - else
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     35:   .table-wrapper
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]     36:     %table.table.inline-table
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99]
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] app/helpers/email_helper.rb:9:in `email_to_canonical_email'
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] app/helpers/email_helper.rb:16:in `email_to_canonical_email_hash'
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] app/models/canonical_email_block.rb:22:in `block in <class:CanonicalEmailBlock>'
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] app/models/canonical_email_block.rb:23:in `block in <class:CanonicalEmailBlock>'
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] app/views/admin/accounts/show.html.haml:33
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] app/controllers/concerns/localized.rb:11:in `set_locale'
May 01 19:29:16 mementomori bundle[3115229]: [4e44a841-18d4-42cb-872c-ce6875253e99] lib/mastodon/rack_middleware.rb:9:in `call'
```

Any idea @renchap?

**Comment by ClearlyClaire** [comment]
Sorry I missed this report. This is indeed a regression from #29305, which actually did not add the feature it claimed to. I'll revert it and then rework it.
